### ISSUE TITLE
Particles: Default for `positionOffset`

### ISSUE THREAD

**Issue by ax3l** [openPMD/openPMD-api#321]
**Is your feature request related to a problem? If so, please describe.**
When writing particle data, we should define the default of `positionOffset` to a *constant record (component)* with default `0` (zero) offset.

This makes it easier for users to write position data if they do not care about fine-vs-coarse precision that `position` + `positionOffset` offer.

Python:
```python
import openPMD


num_cereals = 3

# cereal positions in my breakfast bowl
x = np.arange(num_cereals, dtype=np.double)
y = np.arange(num_cereals, dtype=np.double)

# open file for writing
series = openPMD.Series(
    "cereals.h5",
    openPMD.Access_Type.create
)

cereals = series.iterations[1].particles["cereals"]

datatype = openPMD.Datatype.DOUBLE
extent = [num_cereals]
dataset = openPMD.Dataset(datatype, extent)

cereals["position"]["x"].reset_dataset(dataset)
cereals["position"]["y"].reset_dataset(dataset)

offset = [0]
cereals["position"]["x"].store_chunk(offset, extent, x)
cereals["position"]["y"].store_chunk(offset, extent, y)

# terminate called after throwing an instance of 'std::runtime_error'
# what():  A Record can not be written without any contained RecordComponents: positionOffset
```

**Describe the solution you'd like**

Unless the user sets it explicitly, we should implicitly default in the above example to:
```python
cereals["positionOffset"]["x"].make_constant(0.0)
cereals["positionOffset"]["y"].make_constant(0.0)
```

**Additional context**

Also, this would solve a small bug: if I execute the above example and add the lines
```python
# cereals["positionOffset"]["x"].make_constant(0.0)
cereals["positionOffset"]["y"].make_constant(0.0)
```
the check `A Record can not be written without any contained RecordComponents: positionOffset` does *not throw* anymore, although not every `position` component has a `positionOffset` component counterpart:

```bash
$ h5ls -r cereals.h5
/                        Group
/data                    Group
/data/1                  Group
/data/1/particles        Group
/data/1/particles/cereals Group
/data/1/particles/cereals/position Group
/data/1/particles/cereals/position/x Dataset {3}
/data/1/particles/cereals/position/y Dataset {3}
/data/1/particles/cereals/positionOffset Group
/data/1/particles/cereals/positionOffset/y Group
```

**Comment by anokfireball** [comment]
Full disclosure up front - Including the two [required](https://github.com/openPMD/openPMD-standard/blob/1.1.0/STANDARD.md#records-for-each-particle-species) Records ```position``` and ```positionOffset``` [implicitly](https://github.com/openPMD/openPMD-api/blob/dev/include/openPMD/ParticleSpecies.hpp#L58-L60) did not resonate well with me, but it is the least invasive solution I could come up with. Feel free to propose different solutions if you have any.

> When writing particle data, we should define the default of positionOffset to a constant record (component) with default 0 (zero) offset.

To get this right: You propose to move a check down to the (Particle)```RecordComponent```-level. This check would, for any ```RecordComponent``` ```position/?```, enforce a ```RecordComponent``` ```positionOffset/?```. If this implicit ```positionOffset/?``` does not exist, it will be created and constant-initialized to zero with the same shape as the counterpart?
```python
cereals["positionOffset"]["x"].make_constant(0.0)
```
lacks the datatype and extent of the contant ```RecordComponent```, btw. Both could be implied by the corresponding ```position/?``` (storage requirements are minimal: one scalar and one array attribute).

My initial concern was that users might prematurely ```flush()``` and thus lock themselves out of writing any non-constant ```positionOffset/?```. The benefit of having easier standard-compliance outweighs this, probably.

I'll give it another thought, but this has my general approval.

**Comment by ax3l** [comment]
Yep, but I would not write them on series or user-triggered `flush()` but on series/iteration/species/record shutdown/close, in a sanity check that adds missing defaults as writes.

Maybe we make our life easier for streaming later on, if we add those when user-held references to a certain `Iteration`, `ParticleSpecies`, or `Record` go to zero.

Do we know when a user destructs the handle of an `e` as in
```C++
auto e = o.iterations[1].particles["e"];
```
(and its records and record components in case the user accesses directly through a temporary `ParticleSpecies`)? Is this thinking maybe too tricky?

Maybe always throw on missing and add an explicit `::completeDefaults()`?

If all this implicit automation is too tricky or not expressible, we could also just add nice shortcut helpers and throw as before on missing:

For example,  we could just add a `Record`-level `makeConstant`:
```C++
[...].particles["e"]["positonOffset"].makeConstant({{"x", 0.0}, {"y", 0.0}, {"z", 0.0}});
```
?

Either way, the check (and potentially auto-complete) must be performed exactly when the `ParticleSpecies` gets `flush()`ed (for an iteration), to allow synchronous and asynchronous addition. (That means: also when a whole `Series` or `Iteration` is `flush()`ed, but not when a `Record` or `RecordComponent` is `flush()`ed.)

**Comment by ax3l** [comment]
@C0nsultant wrote:

> Do we know when a user destructs the handle of an `e` as in `auto e = o.iterations[1].particles["e"];` (and it's records)?

Ever since we merged https://github.com/openPMD/openPMD-api/pull/156, that is easily detectable. As everything is `std::shared_ptr`-based, we could simply check for `use_count == 1` (held by the API).
This would mean it's not too tricky, but does not solve the problem and is potentially dangerous. The user might temporarily drop the ref to a `ParticleSpecies` (or any handle for that matter) and then later re-take it from an upper node in the `Series` tree. If we create defaults based on that condition, the same irreversible and potentially undesired file state could be created.

I think the sanest and most helpful way is (as you proposed) to tie this to the lifetime of a `Series`. When that goes out of scope, the user has no way of triggering further flushes in the backend, meaning I/O is done (until the `Series` is manually reopened).
We than can decide whether we want
 - to throw (but only after the file has been closed in the backend so no state is lost) to indicate that the file is *NOT* respecting the standard
 - to 'auto-complete' the `Series` according to the standard

**Comment by ax3l** [comment]
> The user might temporarily drop the ref

Yes, that's also my main concern. As well as direct access through a temporary.

Did you see my edit above (helper record-level `makeConstant` and time and place of checks in the last paragraph of [this](https://github.com/openPMD/openPMD-api/issues/321#issuecomment-466415456))? Overall `Series` is one way or when we `flush()` an iteration's `ParticleSpecies` (something like a transaction).

**Comment by anokfireball** [comment]
> we could just add a `Record`-level `makeConstant`

That does not eliminate the problem of the user forgetting `RecordComponent`s in that call. Also, it is merely syntactic sugar for the individual calls to [`RecordComponent::makeConstant()`](https://github.com/openPMD/openPMD-api/blob/0.7.1-alpha/include/openPMD/RecordComponent.hpp#L108-L109). Edit - I don't even want to start thinking about different datatypes for the constant values in the `std::map< std::string, std::variant< ... > >` that would be required...

> add an explicit `::completeDefaults()`

That is an ideal argument for the `Series::close()` discussed in https://github.com/openPMD/openPMD-api/issues/149.

> the check (and potentially auto-complete) must be performed exactly when the `ParticleSpecies` gets `flush()`ed

Agreed. The check can ultimately be pulled down to the level of a `ParticleSpecies::flush()`. To be very precise, it is the absolute last call to that function in the lifetime of the `Series`. A few preliminary steps (e.g. introducing a `bool Series::m_isClosed`) for #149 can be taken when tackling this issue.
